src/tactic/norm_fin.lean

```diff
@@ -0,0 +1,534 @@
+meta def prove_eq_fin' : expr → expr → expr → expr × expr → expr × expr → eval_fin_m expr
+| n a b (a', pa) (b', pb) :=
+  if a' =ₐ b' then do
+    pure (`(@normalize_fin.eq).mk_app [n, a, b, a', pa, pb])
+  else do
+    (a', pa) ← reduce_fin' ff n a (a', pa),
+    (b', pb) ← reduce_fin' ff n b (b', pb),
+    guard (a' =ₐ b'),
+    pure (`(@normalize_fin.eq).mk_app [n, a, b, a', pa, pb])
+
+/-- Given a function with the type of `prove_eq_fin'`, evaluates it with the given `a` and `b`. -/
+meta def eval_prove_fin
+  (f : expr → expr → expr → expr × expr → expr × expr → eval_fin_m expr)
+  (a b : expr) : tactic expr :=
+do n ← get_fin_type a, eval_fin_m.run $ eval_fin a >>= λ a', eval_fin b >>= f n a b a'
+
+/-- If `a b : fin n`, then `prove_eq_fin a b` proves `a = b`. -/
+meta def prove_eq_fin : expr → expr → tactic expr := eval_prove_fin prove_eq_fin'
+/-- If `a b : fin n`, then `prove_lt_fin a b` proves `a < b`. -/
+meta def prove_lt_fin : expr → expr → tactic expr := eval_prove_fin prove_lt_fin'
+/-- If `a b : fin n`, then `prove_le_fin a b` proves `a ≤ b`. -/
+meta def prove_le_fin : expr → expr → tactic expr := eval_prove_fin prove_le_fin'
+
+section
+open norm_num.match_numeral_result
+
+/-- Given expressions `n` and `m` such that `n` is definitionally equal to `m.succ`, and
+a natural numeral `a`, proves `(b, ⊢ normalize_fin n b a)`, where `n` and `m` are both used
+in the construction of the numeral `b : fin n`. -/
+meta def mk_fin_numeral (n m : expr) : expr → option (expr × expr)
+| a := match match_numeral a with
+  | zero := some (
+    expr.app `(@has_zero.zero (fin %%n)) `(@fin.has_zero %%m),
+    expr.app `(normalize_fin.zero) m)
+  | one := some (
+    expr.app `(@has_one.one (fin %%n)) `(@fin.has_one %%m),
+    expr.app `(normalize_fin.one) m)
+  | bit0 a := do
+    (a', p) ← mk_fin_numeral a,
+    some (`(bit0 %%a' : fin %%n), `(@normalize_fin.bit0).mk_app [n, a', a, p])
+  | bit1 a := do
+    (a', p) ← mk_fin_numeral a,
+    some (
+      `(@_root_.bit1 (fin %%n)).mk_app [`(@fin.has_one %%m), `(@fin.has_add %%n), a'],
+      `(@normalize_fin.bit1).mk_app [m, a', a, p])
+  | _ := none
+  end
+end
+
+/-- The common prep work for the cases in `eval_ineq`. Given inputs `a b : fin n`, it calls
+`f n a' b' na nb` where `a'` and `b'` are the result of `eval_fin` and `na` and `nb` are
+`a' % n` and `b' % n` as natural numbers. -/
+meta def eval_rel {α} (a b : expr)
+  (f : expr → expr × expr → expr × expr → ℕ → ℕ → eval_fin_m α) : tactic α :=
+do n ← get_fin_type a,
+  eval_fin_m.run $ do
+    (nn, n', pn) ← eval_fin_m.eval_n n,
+    (a', pa) ← eval_fin a,
+    (b', pb) ← eval_fin b,
+    na ← eval_fin_m.lift a'.to_nat,
+    nb ← eval_fin_m.lift b'.to_nat,
+    f n (a', pa) (b', pb) (na % nn) (nb % nn)
+
+/-- A `norm_num` extension that evaluates equalities and inequalities on the type `fin n`.
+
+```
+example : (5 : fin 7) = fin.succ (fin.succ 3) := by norm_num
+```
+-/
+@[norm_num] meta def eval_ineq : expr → tactic (expr × expr)
```

Comment:
```suggestion
@[norm_num] meta def eval_rel : expr → tactic (expr × expr)
```
? Since this works on = as well.

Reply:
The function in `norm_num` by the same name also does equality and not equal... Also there's already a function called `eval_rel` in this file